File: LearningLinearAlgebra.Tests/LinearAlgebra/KetTests.cs
using System.Numerics;
using LearningLinearAlgebra.LinearAlgebra.ComplexVectorSpace;
using LearningLinearAlgebra.Numbers;
using FluentAssertions;
using FluentAssertions.Execution;
using Xunit;

namespace LearningLinearAlgebra.Tests.LinearAlgebra;

public class SinglePrecisionCpuKetTests : KetTests<Operator<float>, Bra<float>, Ket<float>, float> { }
public class DoublePrecisionCpuKetTests : KetTests<Operator<double>, Bra<double>, Ket<double>, double> { }

public abstract class KetTests<TOperator, TBra, TKet, TRealNumber>
    where TOperator : IOperator<TOperator, TKet, TBra, TRealNumber>
    where TBra : IBra<TBra, TKet, TRealNumber>
    where TKet : IKet<TKet, TBra, TRealNumber>
    where TRealNumber : IFloatingPointIeee754<TRealNumber>
{
    [Fact]
    public void Dimension_of_the_vector_is_the_number_of_elements_in_basis_vector()
    {
        var ket = TKet.V([(1, 2), (3, 5), (7, 9)]);

        var dimension = TKet.Dimension(ket);

        using var _ = new AssertionScope();

        dimension.Should().Be(3);
        ket.Dimension().Should().Be(TKet.Dimension(ket));
    }

    [Fact]
    public void Addition_is_calculated_as_addition_of_components()
    {
        var a = TKet.V([(1, 2), (3, 5)]);
        var b = TKet.V([(7, 11), (13, 17)]);

        var sum = TKet.Add(a, b);

        using var _ = new AssertionScope();

        sum.Should().BeEquivalentTo(TKet.V([(8, 13), (16, 22)]));
        (a + b).Should().BeEquivalentTo(TKet.Add(a, b));
        a.Add(b).Should().BeEquivalentTo(TKet.Add(a, b));
    }

    [Fact]
    public void Subtraction_is_calculated_as_subtraction_of_components()
    {
        var a = TKet.V([(1, 2), (3, 5)]);
        var b = TKet.V([(7, 11), (13, 17)]);

        var difference = TKet.Subtract(a, b);

        using var _ = new AssertionScope();

        difference.Should().BeEquivalentTo(TKet.V([(-6, -9), (-10, -12)]));
        (a - b).Should().BeEquivalentTo(TKet.Subtract(a, b));
        a.Subtract(b).Should().BeEquivalentTo(TKet.Subtract(a, b));
    }

    [Fact]
    public void Scalar_multiplication_is_calculated_as_scalar_multiplication_of_components()
    {
        var scalar = ComplexNumber<TRealNumber>.C(6, 7);
        var ket = TKet.V([(1, 2), (3, 5)]);

        var product = TKet.Multiply(scalar, ket);

        using var _ = new AssertionScope();

        product.Should().BeEquivalentTo(TKet.V([(-8, 19), (-17, 51)]));
        (scalar * ket).Should().BeEquivalentTo(TKet.Multiply(scalar, ket));
        scalar.Multiply(ket).Should().BeEquivalentTo(TKet.Multiply(scalar, ket));
    }

    [Fact]
    public void Additive_inverse_is_calculated_as_additive_inverse_of_components()
    {
        var ket = TKet.V([(1, 2), (3, 5)]);

        var additiveInverse = TKet.AdditiveInverse(ket);

        using var _ = new AssertionScope();

        additiveInverse.Should().BeEquivalentTo(TKet.V([(-1, -2), (-3, -5)]));
        (-ket).Should().BeEquivalentTo(TKet.AdditiveInverse(ket));
        ket.AdditiveInverse().Should().BeEquivalentTo(TKet.AdditiveInverse(ket));
    }

    [Fact]
    public void Ket_can_be_multiplied_bra_by_applying_the_rules_of_matrix_multiplication()
    {
        var a = TBra.U([(1, 2), (3, 5)]);
        var b = TKet.V([(7, 11), (13, 19)]);

        var product = TKet.Multiply(a, b);

        using var _ = new AssertionScope();

        product.Should().BeEquivalentTo(ComplexNumber<TRealNumber>.C(-71, 147));
        a.Multiply(b).Should().Be(TKet.Multiply(a, b));
        (a * b).Should().BeEquivalentTo(TKet.Multiply(a, b));
    }

    [Fact]
    public void Inner_product_is_a_sum_of_products_of_left_vector_components_and_conjucates_of_right_vector_components()
    {
        var a = TKet.V([(1, 2), (3, 5)]);
        var b = TKet.V([(7, 11), (13, 19)]);

        var innerProduct = TKet.InnerProduct(a, b);

        using var _ = new AssertionScope();

        innerProduct.Should().BeEquivalentTo(ComplexNumber<TRealNumber>.C(163, 11));
        (a * b).Should().BeEquivalentTo(TKet.InnerProduct(a, b));
    }

    [Fact]
    public void Tensor_product_of_vectors_contains_combinations_of_products_of_all_elements_of_both_vectors()
    {
        var a = TKet.V([(1, 2), (3, 5)]);
        var b = TKet.V([(7, 11), (13, 19)]);

        var tensorProduct = TKet.TensorProduct(a, b);

        tensorProduct.Should().BeEquivalentTo(TKet.V([(-15, 25), (-25, 45), (-34, 68), (-56, 122)]));
        a.TensorProduct(b).Should().BeEquivalentTo(TKet.TensorProduct(a, b));
    }

    [Fact]
    public void Conjucate_of_a_vector_is_where_each_element_is_a_complex_conjucate_of_the_original_vector()
    {
        var ket = TKet.V([(1, 2), (3, 5)]);

        var conjucate = TKet.Conjucate(ket);

        using var _ = new AssertionScope();

        conjucate.Should().BeEquivalentTo(TKet.V([(1, -2), (3, -5)]));
        ket.Conjucate().Should().BeEquivalentTo(conjucate);
    }

    [Fact]
    public void Bra_is_a_complex_adjoint_of_ket()
    {
        var a = TKet.V([(1, 2), (3, 5)]);

        var bra = TKet.Bra(a);

        using var _ = new AssertionScope();

        bra.Should().BeEquivalentTo(TBra.U([(1, -2), (3, -5)]));
        a.Bra().Should().BeEquivalentTo(TKet.Bra(a));
    }

    [Fact]
    public void Norm_that_is_the_length_of_the_vector_is_calculated_as_square_root_of_the_inner_product_of_vector_with_itself()
    {
        var ket = TKet.V([(4, 3), (6, -4), (12, -7), (0, 13)]);

        var norm = TKet.Norm(ket);

        using var _ = new AssertionScope();

        norm.Should().Be(RealNumber<TRealNumber>.Sqrt(439));
        ket.Norm().Should().Be(TKet.Norm(ket));
    }

    [Fact]
    public void Distance_of_the_two_vectors_is_the_norm_of_the_difference()
    {

        var a = TKet.V([(1, 2), (3, 5)]);
        var b = TKet.V([(7, 11), (13, 19)]);

        var distance = TKet.Distance(a, b);

        using var _ = new AssertionScope();

        distance.Should().Be(RealNumber<TRealNumber>.Sqrt(413));
        a.Distance(b).Should().Be(TKet.Distance(a, b));
    }

    [Fact]
    public void Normalization_changes_norm_to_one_but_preserves_the_ratio_of_the_components()
    {
        var ket = TKet.V([(1, 2), (3, 5)]);

        var normalized = TKet.Normalized(ket);

        using var _ = new AssertionScope();

        normalized.Should().BeEquivalentTo(TRealNumber.One / ComplexNumber<TRealNumber>.Sqrt(ket * ket) * ket);
        ket.Normalized().Should().BeEquivalentTo(TKet.Normalized(ket));
        normalized.Norm().Round().Should().Be(RealNumber<TRealNumber>.R(1));
    }
}

File: LearningLinearAlgebra.FSharp.Tests/KetTests.fs
namespace LearningLinearAlgebra.LinearAlgebra

open Xunit

module KetTests =

    open LearningLinearAlgebra
    open LearningLinearAlgebra.Numbers.Complex
    open LearningLinearAlgebra.LinearAlgebra.ComplexVectorSpace

    [<Fact>]
    let ``Addition is calculated as addition of components`` () =
        let a = V [| C(1, 2); C(3, 5) |]
        let b = V [| C(7, 11); C(13, 17) |]

        let sum = Ket.Add a b

        sum |> Should.BeEquivalentTo.Complex.Ket(V [| C(8, 13); C(16, 22) |])

    [<Fact>]
    let ``Subtraction is calculated as subtraction of components`` () =
        let a = V [| C(1, 2); C(3, 5) |]
        let b = V [| C(7, 11); C(13, 17) |]

        let sum = Ket.Subtract a b

        sum |> Should.BeEquivalentTo.Complex.Ket(V [| C(-6, -9); C(-10, -12) |])

    [<Fact>]
    let ``Additive inverse is calculated as additive inverse of components`` () =
        let a = V [| C(1, 2); C(3, 5) |]

        let additiveInverse = Ket.AdditiveInverse a

        additiveInverse
        |> Should.BeEquivalentTo.Complex.Ket(V [| C(-1, -2); C(-3, -5) |])

    [<Fact>]
    let ``Scalar multiplication is calculated as scalar multiplication of components`` () =
        let scalar = C(6, 7)
        let ket = V [| C(1, 2); C(3, 5) |]

        let sum = Ket.Multiply scalar ket

        sum |> Should.BeEquivalentTo.Complex.Ket(V [| C(-8, 19); C(-17, 51) |])

    [<Fact>]
    let ``Scalar multiplications by real number is calculated as scalar multiplication of components`` () =
        let scalar = 6.0
        let ket = V [| C(1, 2); C(3, 5) |]

        let sum = Ket.MultiplyByReal scalar ket

        sum |> Should.BeEquivalentTo.Complex.Ket(V [| C(6, 12); C(18, 30) |])

    [<Fact>]
    let ``Ket can be multiplied bra by applying the rules of matrix multiplication`` () =
        let bra = U [| C(1, 2); C(3, 5) |]
        let ket = V [| C(7, 11); C(13, 19) |]

        let product = Ket.MultiplyByBra bra ket

        product |> Should.BeEquivalentTo.Complex.Number(C(-71.0, 147))

    [<Fact>]
    let ``Inner product is a sum of products of left vector components and conjucates of right vector components`` () =
        let a = V [| C(1, 2); C(3, 5) |]
        let b = V [| C(7, 11); C(13, 19) |]

        let innerProduct = Ket.InnerProduct a b

        innerProduct |> Should.BeEquivalentTo.Complex.Number(C(163, 11))

    [<Fact>]
    let ``Conjucate of a vector is where each element is a complex conjucate of the original vector`` () =
        let vector = V [| C(1, 2); C(3, 5) |]

        let conjucate = Ket.Conjucate vector

        conjucate |> Should.BeEquivalentTo.Complex.Ket(V [| C(1, -2); C(3, -5) |])

    [<Fact>]
    let ``Bra is a complex adjoint of ket`` () =
        let a = V([| C(1, 2); C(3, 5) |])

        let bra = Ket.Bra(a)

        bra |> Should.BeEquivalentTo.Complex.Bra(U([| C(1, -2); C(3, -5) |]))

    [<Fact>]
    let ``Norm that is the length of the vector is calculated as square root of the inner product of vector with itself``
        ()
        =
        let ket = V([| C(4, 3); C(6, -4); C(12, -7); C(0, 13) |])

        let norm = Ket.Norm ket

        norm |> Should.BeEquivalentTo.Real.Number(sqrt (439.0))

    [<Fact>]
    let ``Distance of the two vectors is the norm of the difference`` () =
        let a = V([| C(1, 2); C(3, 5) |])
        let b = V([| C(7, 11); C(13, 19) |])

        let distance = Ket.Distance a b

        distance |> Should.BeEquivalentTo.Real.Number(sqrt (413.0))

    [<Fact>]
    let ``Normalization changes norm to one but preserves the ratio of the components`` () =
        let ket = V([| C(1, 2); C(3, 5) |])

        let normalized = Ket.Normalized ket

        normalized
        |> Should.BeEquivalentTo.Complex.Ket(
            V(
                [| C(0.16012815380508713, 0.32025630761017426)
                   C(0.48038446141526137, 0.8006407690254357) |]
            )
        )

        Ket.Norm(Ket.Normalized(ket)) |> Should.BeEquivalentTo.Real.Number(1)
